Automattic/WP-Job-Manager
# Change how `[jobs]` maintains state when navigating away and refreshing

Fixes #1712 
Fixes #1680 

This fixes a lot of strange behavior with the `[jobs]` shortcode. Since it is a pretty major refactor, I also ran `prettier` on the file. Eventually I'd like to run this across all JS files in WPJM.

#### Changes proposed in this Pull Request:

* Adds back in the loading indicator.
* Shows number of results and allows for reset/rss if the only filter change was deselecting a job type.
* Properly sets up the job type checkboxes after refresh or going back in browser history.
* Moves from HTML5's History API to SessionStorage. No more weird hash (#s=1) in URL and allows for multiple forms on same page.

Key changes to behavior:
* Within a browser session, always persist the form values on both refresh and going back to the `[jobs]` page after navigating away.
* Within a browser session, use cached/previously retrieved results when going back to `[jobs]` page _after navigating to a job listing_. If a user clicks elsewhere in the page and then goes back or just refreshes, the form will persist but the results will be reloaded _starting on page 1_. @alexsanford and I talked about this for a while and it seemed to make the most sense. 

#### Caveats/Known Issues

- If you have the Job Tags plugin activated, the tags filter still won't persist. That change will need to be made in the Job Tags plugin. There  is already an open issue for this.
- If you open a job listing in a new tab, the first time you refresh the `[jobs]` page it will persist the results. Refresh again and it will clear those cached results and start back on page 1.

#### Testing instructions:

* Create multiple job listings and make sure they are published. I'd recommend at least 10.
* Create a pages with:
  - `[jobs]`
  - `[jobs per_page="3"]` (change arg) 
  - `[jobs show_pagination="true"]`
  - `[jobs per_page="3" show_filters="false"]` (form not there to persist but results/page should per above)
  - `[jobs per_page="3" show_pagination="true" show_filters="false"]` (form not there to persist but results/page should per above)
* Test to make sure new behavior works across browsers and doesn't break major themes.
